### PR TITLE
Refs #25809 - handle OIDC JWT in apache service

### PR DESCRIPTION
**Pull request by rabajaj0509**

### CODE REVIEW

**Comment by rabajaj0509** [comment]
@tbrisker can you please take a look at this PR? test cases will follow shortly.

File: app/services/sso/openid_connect.rb
Comment (tbrisker) [inline app/services/sso/openid_connect.rb]: how is HTTP_OIDC_ACCESS_TOKEN header set? is this the default setting by keycloak sso or is it some header we define on the keycloak server?
In other words - is it possible to instead just set the authorization header like api requests and use the same flow?

Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb]: @tbrisker 

> how is HTTP_OIDC_ACCESS_TOKEN header set? is this the default setting by keycloak sso or is it some header we define on the keycloak server?

The HTTP_OIDC_ACCESS_TOKEN is set by the `mod_auth_openidc` apache module. This is therefore a header that is inserted by the module and passed on to Foreman.

> In other words - is it possible to instead just set the authorization header like api requests and use the same flow?

Imo, that would not be the case.

Let me know your thoughts.

Comment (tbrisker) [reply app/services/sso/openid_connect.rb]: Ok, thanks for the clarification.

**Comment by tbrisker** [comment]
This seems reasonable but I'm not very familiar with sso/apache auth modules. @ares @ekohl maybe you can take a look? (pending additional tests added to verify this)

File: app/services/sso/openid_connect.rb
Comment (ekohl) [reply app/services/sso/openid_connect.rb]: If that module is not present, any user can present it. It looks like the token is still validated regardless of how it came in, so that isn't a security concern, right?

> The HTTP_OIDC_ACCESS_TOKEN is set by the `mod_auth_openidc` apache module. This is therefore a header that is inserted by the module and passed on to Foreman.

That implies an installer change. Will this be required?

Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb]: > If that module is not present, any user can present it. It looks like the token is still validated regardless of how it came in, so that isn't a security concern, right?

@ekohl 

No, imo, there should not be a security issue because our validation checks each section(header, payload and signature) of the JWT and only a intended token would able to create/login with a token. 
> 
> > The HTTP_OIDC_ACCESS_TOKEN is set by the `mod_auth_openidc` apache module. This is therefore a header that is inserted by the module and passed on to Foreman.
> 
> That implies an installer change. Will this be required?

 I doubt that, earlier folks were using SAML to do the authentication(https://www.theforeman.org/2018/06/using-saml-for-single-sign-on-to-foreman-through-keycloak.html) but installer would remain unaffected. Only difference would be that in case of SAML, the variables could be manipulated in the way we wanted them, eg: `REMOTE_USER`, could be passed whereas here we are receiving a JWT.

Let me know what you think about the installer changes.

Comment (rabajaj0509) [reply app/services/sso/openid_connect.rb]: @ekohl on second thought, should we also have a installer change that installs `mod_auth_openidc` for Foreman? some option like `--enable-sso`.

What do you think?

File: app/services/sso/apache.rb
Comment (ares) [inline app/services/sso/apache.rb]: this could get out of sync with https://github.com/theforeman/foreman/blob/ef511e1c58c528e073a55c021a0c257d0b73b012/app/services/sso/openid_connect.rb#L50

you can extract this constant to https://github.com/theforeman/foreman/blob/ef511e1c58c528e073a55c021a0c257d0b73b012/app/services/sso/base.rb#L2 and reuse on both places I guess

File: test/unit/sso/apache_test.rb
Comment (ares) [inline test/unit/sso/apache_test.rb]: isn't this still disabled by line 31? `Setting["authorize_login_delegation_api"]` didn't get reset before this test

**Review by ares — CHANGES_REQUESTED** [review]
looks reasonable, left few small comments

**Comment by ares** [comment]
LGTM, waiting on ack from Nikhil, that it works

**Review by ares — CHANGES_REQUESTED** [review]
Test failures related

**Review by ares — COMMENTED** [review]
Code-wise ready, any update on user testing?

**Review by ntkathole — APPROVED** [review]
ACK. Login via SSO works.

Although, the functionality is working there are a few issues that I want to point out:
1) Show error on invalid oidc issuer 
2) Token does not expire on logout

I don't want to block this PR so raised these issues in redmine to track https://projects.theforeman.org/issues/28087 and https://projects.theforeman.org/issues/28086.


ezgif com-video-to-gif

**Review by ares — APPROVED** [review]
Thanks @rahulbajaj0509 and @ntkathole, I assume Rahul will take a look at the reported issues. Katello pulp3 issues are unrelated, hence nothing blocks this PR to get in. Merging!

**Pull request merged**
